[CK_BUILDER] convolution testing (#3267)
ROCm/composable_kernel: experimental/builder/include/ck_tile/builder/testing/README.md

Reviewer: This reads very well is is straightforward. Very nice work!

# CK-Builder Testing Utilities

This directory contains testing utilities designed to simplify the process of writing unit tests for GPU kernels built with `ck_tile::builder`. These utilities enable a clean, expressive **Given-When-Then** (Given-When-Then) testing pattern that separates test setup, execution, and validation.

> ## Overview

Reviewer 2: Refer back to CK builder documentation, see Ville's example in https://github.com/ROCm/composable_kernel/blob/develop/experimental/builder/include/ck_tile/builder/factory/README.md.

Author: I've also updated the main file to link to the different modules.

    Transfer transfer;

    // Additional tuning parameters
    // ...
};
static_assert(ck_tile::builder::ConvAlgorithmDescriptor<ConvAlgorithm>);
constexpr auto ALGORITHM = ConvAlgorithm{
    .thread_block = /* ... */;
    .gridwise_gem = /* ... */;
    .transfer = /* ... */;
    // ...
};
```

#### Building the Kernel

The `Builder` combines the signature (what to compute) with the algorithm (how to compute it) to generate a kernel type which represents the operation. The implementation details, including invocation method, depend on the particular signature and algorithm.

```cpp
using Conv = ck_tile::builder::ConvBuilder<SIGNATURE, ALGORITHM>::Instance;
auto conv = Conv{};
```

#### Invoking the Kernel

After creating the kernel instance, it can be invoked by passing the instance, the arguments, the inputs, and the outputs to `run()`. This operation writes results into the buffers in `outputs`.

```cpp
ck_tile::test::run(conv, args, inputs.get(), outputs.get());
```

### Then: Verifying the Results

The "Then" phase validates that the kernel produced the expected output. This is done by running a reference kernel and comparing the results.

#### Building the Reference Kernel

> The reference kernel is just another kernel instance of the builder, one that's been externally verified to produce the correct results. As this kernel is also running on the GPU, we can use it to perform tests far more quickly than when comparing the outputs to a CPU-based reference implementation.

Reviewer: I really like this approach!
I'm sure you've already thought of it this, but just in case not, we should be adding automated checks to verify the known reference kernels are consistent over time.

Author: Yep, the reference kernel should be "known good" in the context of the CK Builder test infrastructure. I believe that the current implementation already has this.

};
static_assert(ck_tile::builder::ConvAlgorithmDescriptor<ReferenceAlgorithm>);
constexpr auto REFERENCE_ALGORITHM = ReferenceAlgorithm{
    .specialization = ck_tile::builder::ConvAlgorithmSpecialization::REFERENCE;
};
using ReferenceConv = ck_tile::builder::ConvBuilder<SIGNATURE, REFERENCE_ALGORITHM>::Instance;
auto reference_conv = ReferenceConv{};
```

This instance can then be invoked using `ck_tile::testing::run()`, the same as the kernel to be tested. Note that another instance of the `Outputs` structure needs to be passed here in order to store the results.

```cpp
auto reference_outputs = ck_tile::test::allocate_outputs(args);
ck_tile::test::run(conv, args, inputs.get(), reference_outputs.get());
```

#### `Validator<SIGNATURE>`

The `Validator` class encapsulates the validation logic. It performs on-GPU correctness checks by comparing two instances of the `Outputs` structure.

```cpp
ck_tile::test::Validator<SIGNATURE> validator(outputs.get(), reference_outputs.get());
```

The `Validator` provides methods that return GoogleMock matchers, enabling clean integration with GoogleTest:

```cpp
EXPECT_THAT(validator.result(), validator.is_ok());
```

> The `is_ok()` matcher checks that the output is numerically correct within acceptable tolerances. The `Validator` can also provide more detailed diagnostics, such as:

Reviewer: I don't see the custom matchers existing yet, but just recommending the name of the is_ok() matcher should convey what is actually being checked. Maybe something like matches_reference_output(), is_within_tolerance(), is_numerically_correct() for example.
```suggestion
The `matches_reference_output()` matcher checks that the output is numerically correct within acceptable tolerances. The `Validator` can also provide more detailed diagnostics, such as:
```

Author: I think that the details of this will change indeed. Ive updated the docs to your suggestion for now.

See the [main builder documentation](../README.md) for an overview.

## Overview

Testing GPU kernels typically involves significant boilerplate: allocating device memory, initializing test data, launching kernels, and validating results. The utilities in this directory abstract away these repetitive tasks, allowing you to focus on defining test cases and verifying correctness.

The core components are:

- **`Args`**: A struct template that holds runtime parameters for a specific test case.
- **`Input`** and **`Output`**: Helper classes that groups operation inputs and outputs.
- **`Validator`**: A utility that performs on-GPU validation and integrates with GoogleTest/GoogleMock.

Together, these components enable a structured approach to kernel testing that mirrors the Given-When-Then pattern commonly used in behavior-driven development.

## The Given-When-Then Testing Pattern

The Given-When-Then pattern organizes tests into three distinct phases:

1. **Given**: Set up the preconditions and test data
2. **When**: Execute the action being tested
3. **Then**: Verify the expected outcome

> 

Reviewer 3: I'd move this up personally. Make it clear what it is from the get-go.

Author: Hmm, I'm not sure that I like that. Right now the pattern and its details fits nicely into its own section, and moving the explanation of the given-then-when pattern up starts the "implementation details" of the readme to soon imo. I've left it like this for now.

};
```

#### Run-time Arguments

The `Args` struct template provides the **runtime parameters** for your test case. It is parameterized by the `SIGNATURE` and contains fields for tensor dimensions, strides, dilations, and other dynamic properties. Note that the exact parameters required for each `Args` depends on the `SIGNATURE`: For example, a `SIGNATURE` that represents a forward convolution requires specifying the number of batches, groups, input- and output-channels, filter dimensions, filter strides, and so on. A `SIGNATURE` that represents a simple GEMM operation may instead require only the dimensions of the A-, B- and C-matrices.

```cpp
ck_tile::test::Args<SIGNATURE> args = {
    .lengths = {
        .batch_size      = 128,
        .groups          = 1,
        .input_channels  = 64,
        .output_channels = 128,
        .image           = {.height = 56, .width = 56},
        .filter          = {.height = 3,  .width = 3},
    },
    .filter_strides  = {.height = 1, .width  = 1},
    .filter_dilation = {.height = 1, .width  = 1},
    .input_left_pad  = {.width  = 1, .height = 1},
    .input_right_pad = {.width  = 1, .height = 1},
};
```

#### Tensor Memory Management

> Tensor memory is passed around using the `Inputs<SIGNATURE>` and `Outputs<SIGNATURE>` structures. These group all inputs and outputs for an operation. Note that these structures do not "own" the memory inside: They only logically group the inputs so that they can be passed around as a common type. The amount of inputs and outputs may differ depending on the `SIGNATURE`, and this avoids having to pass around additional values and accept additional parameters in those situations.

Reviewer 3: ```suggestion
Tensor memory is passed using the `Inputs<SIGNATURE>` and `Outputs<SIGNATURE>` structures. These group all inputs and outputs for an operation. Note that these structures do not "own" the memory inside: They only logically group the inputs so that they can be passed as a common type. The amount of inputs and outputs may differ depending on the `SIGNATURE`, and this avoids having to pass additional values and accept additional parameters in those situations.
```

"passed around" sounds like a tray of hors d'oeuvres at a party :)

It would be good to say what the memory is being passed to/from.

The exact fields in `Inputs` and `Outputs` depend again on the particular `SIGNATURE` that they are constructed with. In general, these structures are intended to be freely constructible from external data and only serve to group relevant information. Automatic memory management can be performed using the `UniqueInputs<SIGNATURE>` and `UniqueOutputs<SIGNATURE>` structures instead. The `alloc_inputs` and `alloc_outputs` functions are used to initialize these types: They take an `Args` structure and allocate the appropriate amounts of memory. `.get()` is used to return an instance of the appropriate `Input` or `Output`.

```cpp
auto inputs = ck_tile::test::allocate_inputs(args);
auto outputs = ck_tile::test::allocate_outputs(args);
```

> Note that these functions merely _allocate_ memory: After allocation, the memory is yet uninitialized.

Reviewer 3: ```suggestion
Note that these functions merely _allocate_ memory: After allocation, the memory is still uninitialized.
```
